src/Stratis.Bitcoin.Features.Consensus/ProvenBlockHeaders/ProvenBlockHeaderStore.cs

```diff
@@ -288,43 +310,134 @@ private async Task SaveAsync()
                 this.pendingTipHashHeight = null;
             }
 
-            this.CheckItemsAreInSequence(pendingBatch.Keys.ToList());
+            this.CheckItemsAreInConsecutiveSequence(pendingBatch.Keys.ToList());
 
+            // Save the items to disk.
             using (new StopwatchDisposable(o => this.performanceCounter.AddInsertTime(o)))
             {
                 // Save the items to disk.
-                await this.provenBlockHeaderRepository.PutAsync(pendingBatch.Values.ToList(), hashHeight);
+                await this.provenBlockHeaderRepository.PutAsync(pendingBatch.Values.ToList(), hashHeight).ConfigureAwait(false);
+
+                this.TipHashHeight = this.provenBlockHeaderRepository.TipHashHeight;
+            }
+        }
+
+        /// <summary>
+        /// Will recover the <see cref="IProvenBlockHeaderStore"/> tip to the most recent <see cref="ProvenBlockHeader"/> 
+        /// that exists in the <see cref="IProvenBlockHeaderRepository"/> and/or <see cref="ChainedHeader"/>.
+        /// </summary>
+        /// <param name="newChainedHeader">Current <see cref="ChainedHeader"/> tip with all its ancestors.</param>
+        /// <exception cref="ProvenBlockHeaderException">
+        /// Thrown when :
+        /// <list type="bullet">
+        /// <item>
+        /// <term>Corrupt.</term>
+        /// <description>When the latest <see cref="ProvenBlockHeader"/> does not exist in the <see cref="BlockHeaderRepository"/>.</description>
+        /// </item>
+        /// <item>
+        /// <term>No common block hash exists.</term>
+        /// <description>When the chain header block hash cannot be found within the <see cref="ProvenBlockHeaderStore"/>.</description>
+        /// </item>
+        /// <item>
+        /// <term><see cref="ChainedHeader"/> ahead.</term>
+        /// <description>When the <see cref="newChainedHeader"/> tip is ahead of the <see cref="ProvenBlockHeaderStore"/> tip.</description>
+        /// </item>
+        /// </list>
+        /// </exception>
+        private async Task RecoverStoreTipAsync(ChainedHeader newChainedHeader)
+        {
+            int tipHeight = this.provenBlockHeaderRepository.TipHashHeight.Height;
+
+            ProvenBlockHeader latestHeader = await this.provenBlockHeaderRepository.GetAsync(tipHeight);
+
+            if ((tipHeight > 0) && (latestHeader == null))
+            {
+                // Happens when the proven header store is corrupt.
+                throw new ProvenBlockHeaderException("Proven block header store failed to recover.");
+            }
 
-                this.TipHashHeight = hashHeight;
+            // This happens when the new chain reorg is behind this current chain store.
+            if (newChainedHeader.Height < tipHeight)
```

Comment:
From line 361 to line 385. Why do we need to do this? 
We can just get proven header at the height of consensus tip and set it as a tip

Reply:
Ah @noescape00 makes a good point, because we override entries in disk we just need to set PH store tip based on `newChainedHeader`.

so all you need to do is repo.getHeight(newChainedHeader.Height) and set this as store tip

but you still need to throw if PH tip is bellow `newChainedHeader`